File: CourseSchedule/CourseSchedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = [[] for _ in xrange(numCourses)]
        visited = [0 for _ in xrange(numCourses)]
        for i in prerequisites:
            graph[i[0]].append(i[1])
        for i in xrange(numCourses):
            if not crourseDFS(graph, visited, i):
                return False
        return True

def crourseDFS(graph, visited, course):
    if visited[course] == -1:
        return False
    if visited[course] == 1:
        return True
    visited[course] = -1
    for i in graph[course]:
        if not crourseDFS(graph, visited, i):
            return False
    visited[course] = 1
    return True

File: CourseSchedule/CourseSchedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<vector<int>> graph(numCourses, vector<int>(0));
        vector<int> visited(numCourses, 0);
        for(auto &i : prerequisites) {
        	graph[i.first].push_back(i.second);
        }
        for(int i = 0; i < numCourses; i++) {
        	if(!crourseDFS(graph, visited, i))
        		return false;
        }
        return true;
    }
private:
	bool crourseDFS(vector<vector<int>> &graph, vector<int> &visited, int i) {
		if(visited[i] == -1)
			return false;
		if(visited[i] == 1)
			return true;
		visited[i] = -1;
		for(auto j : graph[i]) {
			if(!crourseDFS(graph, visited, j))
				return false;
		}
		visited[i] = 1;
		return true;
	}
};